File: Backend/src/main/java/com/ader/backend/entity/Offer.java
package com.ader.backend.entity;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;

import javax.persistence.*;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;

@EqualsAndHashCode(callSuper = true)
@Entity
@Table(name = "offers")
@Data
@NoArgsConstructor
@SequenceGenerator(name = "generic_gen", sequenceName = "offer_seq", allocationSize = 1)
public class Offer extends BaseEntity {

    @Column
    private String name;

    @Column
    private String description;

    @Column
    private Timestamp expireDate;

    @ManyToOne
    @JoinColumn
    @ToString.Exclude
    private User author;

    @ManyToOne
    @JoinColumn
    @ToString.Exclude
    private User assignee;

    @OneToMany(mappedBy = "offer", cascade = CascadeType.ALL)
    @ToString.Exclude
    private List<Bid> bids = new ArrayList<>();

    @ManyToMany
    @JoinTable(
            name = "offer_category",
            joinColumns = {
                    @JoinColumn(name = "offer_id", referencedColumnName = "id"),
            },
            inverseJoinColumns = {
                    @JoinColumn(name = "category_id", referencedColumnName = "id")
            }
    )
    private List<Category> categories = new ArrayList<>();

    @ToString.Exclude
    @OneToMany(mappedBy = "offer", cascade = CascadeType.ALL)
    private List<File> files = new ArrayList<>();

    @Enumerated(EnumType.STRING)
    private OfferStatus offerStatus = OfferStatus.OPEN;

    @Enumerated(EnumType.STRING)
    private Status status = Status.ACTIVE;
}

File: Frontend/src/main/web/src/app/model/offer/offer.ts
import {Status} from "../status/status.enum";
import {BidViewModel} from "../bid/bid-view-model";
import {CategoryViewModel} from "../category/category-view-model";
import {OfferMediaViewModel} from "../offermedia/offer-media-view-model";
import {OfferStatus} from "../offerstatus/offer-status.enum";
import DateTimeFormat = Intl.DateTimeFormat;

export interface Offer {
  id: number;
  name: string;
  description: string;
  authorName: string;
  assigneeName: string;
  bids: BidViewModel[];
  categories: CategoryViewModel[];
  offerMedia: OfferMediaViewModel[];
  offerStatus: OfferStatus;
  expireDate: DateTimeFormat;
  created: DateTimeFormat;
  updated: DateTimeFormat;
  status: Status;
}
